### DISCUSSION TITLE
PorousFlow: NeumannBC not affecting Pressure Gradient.

### DISCUSSION THREAD

**Discussion by batodon** [idaholab/moose#22918] [Q&A General] [answered]
Hi,
I’m using the PorousFlow module to perform a simple steady-state darcy flow through a unit square model with displacement BC of 1e4 at the top and zero everywhere. The inlet has a `NeumannBC` of `10e-10kg/s`, and the outlet has a `pressure` of `1MPa`. After running the simulation and comparing the results with the benchmark study, I noticed that my pressure gradient was extremely low (also inaccurate). I’m guessing the way I applied the NeumannBC on the inlet is not the best because the pressure gradient only changes when I change the pressure at the outer boundary (i.e., the legend shows the pressure gradient from the initial pressure condition to the outer pressure condition). Could someone please help? I attached my input file, an image of my result, and the benchmark. Thanks. 

Benchmark result:
![Benchmark_result](https://user-images.githubusercontent.com/113298585/206512516-4be20f92-2281-46dc-aa26-1414b7c56742.PNG)
MOOSE result:
![my_moose_result](https://user-images.githubusercontent.com/113298585/206512564-80e26a0b-6d0e-429a-97f0-1488d6df2392.PNG)

```


[Mesh]
  [efmcube]
    type = FileMeshGenerator
    file = square.inp
 []
[]


[GlobalParams]
 displacements = 'disp_x disp_y'
  gravity = '0 0 0'
  biot_coefficient = 0.0
  PorousFlowDictator = dictator
[]


[UserObjects]
  [dictator]
    type = PorousFlowDictator
    porous_flow_vars = 'pwater disp_x disp_y '
    number_fluid_phases = 1
    number_fluid_components = 1
  []
  [pc]
    type = PorousFlowCapillaryPressureVG
    alpha = 1E-6
   m = 0.6
  []
[]


[Variables]
  [pwater]
    initial_condition = 1.01325e5 
  []
  [disp_x]
    scaling = 1E-5
  []
  [disp_y]
    scaling = 1E-5
  []
[]


[AuxVariables]
  [effective_fluid_pressure]
    family = MONOMIAL
    order = CONSTANT
  []
    [darcy_vel_x]
    family = MONOMIAL
    order = CONSTANT
  []
  [darcy_vel_y]
    family = MONOMIAL
    order = CONSTANT
  []
  [stress_xx]
    family = MONOMIAL
    order = CONSTANT
  []
  [stress_yy]
    family = MONOMIAL
    order = CONSTANT
  []
  [porosity]
    family = MONOMIAL
    order = CONSTANT
  []
    [permeability]
    family = MONOMIAL
    order = CONSTANT
  []
    [strain_normal]
    family = MONOMIAL
    order = CONSTANT
  []
[]

[AuxKernels]
  [effective_fluid_pressure]
    type = ParsedAux
    args = 'pwater'
    function = 'pwater'
    variable = effective_fluid_pressure
  []
    [darcy_vel_x]
    type = PorousFlowDarcyVelocityComponent
    component = x
    variable = darcy_vel_x
    fluid_phase = 0                             # OPTIONAL for single-phase
    execute_on = TIMESTEP_END
  []
  [darcy_vel_y]
    type = PorousFlowDarcyVelocityComponent
    component = y
    variable = darcy_vel_y
    fluid_phase = 0                             # OPTIONAL for single-phase
    execute_on = TIMESTEP_END
  []
  [stress_xx]
    type = RankTwoScalarAux
    variable = stress_xx
    rank_two_tensor = stress
    scalar_type = MinPrincipal
    point1 = '0 0 0'
    point2 = '0 0 1'
    execute_on = timestep_end
  []
  [stress_yy]
    type = RankTwoScalarAux
    variable = stress_yy
    rank_two_tensor = stress
    scalar_type = MidPrincipal
    point1 = '0 0 0'
    point2 = '0 0 1'
    execute_on = timestep_end
  []
  [porosity]
    type = PorousFlowPropertyAux
    variable = porosity
    property = porosity
    execute_on = timestep_end
  []
    [permeability]
    type = PorousFlowPropertyAux
    variable = permeability
    property = permeability
    execute_on = timestep_end
  []
  [strain_normal]
    type = MaterialRealAux
    variable = strain_normal
    property = fracture_normal_strain_qp
    execute_on = timestep_end
  []
[]


[Kernels]
  [flux_water]
    type = PorousFlowAdvectiveFlux
    variable = pwater
    use_displaced_mesh = false
  []
  [grad_stress_x] 
    type = StressDivergenceTensors 
    variable = disp_x
    use_displaced_mesh = false
    component = 0
  []
  [poro_x]
    type = PorousFlowEffectiveStressCoupling
    variable = disp_x
    use_displaced_mesh = false
    component = 0
  []
  [grad_stress_y]
    type = StressDivergenceTensors
    variable = disp_y
    use_displaced_mesh = false
    component = 1
  []
  [poro_y]
    type = PorousFlowEffectiveStressCoupling
    variable = disp_y
    use_displaced_mesh = false
    component = 1
  []
[]

#[Modules]
#  [FluidProperties]
#    [true_water]
#      type = Water97FluidProperties
#    []
#    [tabulated_water]
#      type = TabulatedFluidProperties
#      fp = true_water
#      temperature_min = 275
#      pressure_max = 1E8
#      interpolated_properties = 'density viscosity enthalpy internal_energy'
#      fluid_property_file = water97_tabulated_11.csv
#    []
#   []
#[]


[Modules]
  [FluidProperties]
    [water]
      type = SimpleFluidProperties
      bulk_modulus = 2.27e14
      density0 = 2.15e3
      viscosity = 1e-3
    []
  []
[]


[Materials]
  [temperature]
    type = PorousFlowTemperature
    temperature = 313.15
    use_displaced_mesh = false
  []
  [saturation_calculator]
    type = PorousFlow1PhaseP
    porepressure = pwater
    capillary_pressure = pc
  []
    [massfrac]
    type = PorousFlowMassFraction
  []
  [water_viscosity_density]
    type = PorousFlowSingleComponentFluid
   # fp = tabulated_water                       #viscosity and density obtained from EOS    
    fp = water
    phase = 0
  []
   [porosity_mat]
    type = PorousFlowPorosity
    fluid = true
    mechanical = true
    thermal = true
    porosity_zero = 0.0
    reference_temperature = 330
    reference_porepressure = 20E6
    thermal_expansion_coeff = 15E-6 # volumetric
    solid_bulk = 8E9 # unimportant since biot = 1
  []
    [permeability]
    type = PorousFlowEmbeddedFracturePermeabilityBase
    a =  0.01
    b0 = 0.0
    e0 = 1e-5
    km = 1e-20
    rad_xy = 0.785398
    rad_yz = 0.785398
    jf = 1
    n = "0 1 0"

    # type = PorousFlowPermeabilityKozenyCarman
    # poroperm_function = kozeny_carman_phi0
    # phi0 = 0.1
    # n = 2
    #m = 2
    #k0 = 1E-12

  []
  [relperm_water]
    type = PorousFlowRelativePermeabilityCorey
    n = 4.0
    s_res = 0.1
    sum_s_res = 0.2
    phase = 0
  []

  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1E9
    poissons_ratio = 0.3
  []
  [strain]
    type = ComputeSmallStrain
    eigenstrain_names = 'initial_stress'
  []
  [initial_strain]
    type = ComputeEigenstrainFromInitialStress
   initial_stress = '20E6 0 0  0 20E6 0  0 0 20E6'
    eigenstrain_name = initial_stress
  []
  [stress]
    type = ComputeLinearElasticStress
  []

  [effective_fluid_pressure_mat]
    type = PorousFlowEffectiveFluidPressure
  []
  [volumetric_strain]
    type = PorousFlowVolumetricStrain
  []
[]

[BCs]
  [u_fix_left_x]
    type = DirichletBC
    variable = disp_x
    value = 0
    boundary = 'inlet'       
  []
  [u_fix_top_y]
    type = DirichletBC
    variable = disp_y
    value = 1.1e-4
    boundary = 'top'         
  []
  [u_fix_bottom_y]
    type = DirichletBC
   variable = disp_y
   value = 0
    boundary = 'bottom'      
  []
  [u_fix_right_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'outlet' 
    value = 0      
  []
   [pressure_right_x]
    type = DirichletBC
    boundary = 'outlet'              
    variable = pwater
    value = 1e5
    use_displaced_mesh = false 
  []
    [fluid_pressure_left]
    type = Pressure
    boundary = inlet
    variable = disp_x
    component = 0
   postprocessor = constrained_effective_fluid_pressure_at_wellbore
   use_displaced_mesh = false
  []
  [flux_left]
   type = PorousFlowSink
    boundary = inlet
    variable = pwater
    fluid_phase = 0
    flux_function = -10e-10
    use_displaced_mesh = false
  []
#[./left]
#  type = NeumannBC
#  variable = disp_x
#  boundary = inlet
#  value = -10e-10
#[../]
[]

[Postprocessors]
  [effective_fluid_pressure_at_wellbore]
    type = PointValue
    variable = effective_fluid_pressure
    point = '1 0 0'
    execute_on = timestep_begin
    use_displaced_mesh = false
  []
  [constrained_effective_fluid_pressure_at_wellbore]
    type = FunctionValuePostprocessor
    function = constrain_effective_fluid_pressure
    execute_on = timestep_begin
 []
   [fracture_normal_strain]
    type = PointValue
    variable = strain_normal
    point = '1 0 0'
    execute_on = timestep_begin
  []
   [pwater_pressure]
    type = PointValue
    variable = pwater
    point = '1 0 0'
    execute_on = timestep_begin
  [] 
   [permeability]
    type = PointValue
    variable = permeability
    point = '1 0 0'
    execute_on = timestep_begin
  []
[]

[Functions]
  [constrain_effective_fluid_pressure]
    type = ParsedFunction
    vars = effective_fluid_pressure_at_wellbore
    vals = effective_fluid_pressure_at_wellbore
    value = 'max(effective_fluid_pressure_at_wellbore, 1.01325e5)'
  []
[]

[Preconditioning]
  active = basic
  [basic]
    type = SMP
    full = true
    petsc_options = '-ksp_diagonal_scale -ksp_diagonal_scale_fix'
    petsc_options_iname = '-pc_type -sub_pc_type -sub_pc_factor_shift_type -pc_asm_overlap'
    petsc_options_value = ' asm      lu           NONZERO                   2'
  []
  [preferred_but_might_not_be_installed]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
    petsc_options_value = ' lu       mumps'
  []
[]

[Executioner]
  type = Steady
  solve_type = Newton
  [TimeStepper]
    type = IterationAdaptiveDT
    dt = 1E3  # 1E3 1E4
    growth_factor = 1.2
    optimal_iterations = 10
  []
  nl_abs_tol = 1E-1
[]

[Outputs]
  exodus = true
  [csv]
  type = CSV
  execute_on = 'initial timestep_end'
  []
[]
```

**Comment by cpgr** [comment]
Possibly just a typo? You're flux in the BC's on the inlet is `10e-10` in the input file, so basically zero. Is that really the flux?

**Reply by batodon** [reply]
Hi @cpgr . Sorry, that's a typo. The BC is `10e-10kg/s` (yeah, i guess basically zero) in the published article. Even if I changed it to `10e10kg/s`, I got the following error:

```
Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 50
 Solve Did NOT Converge!
Aborting as solve did not converge
```

**Reply by cpgr** [reply]
Yeah, I wouldn't imagine it would work with 10e10 kg/s!

The `flux_function` in `PorousFlowSink` is measured in `kg/m^2/s` as it is integrated over the boundary. Unless your `inlet` boundary has an area of `1m^2` then your flux isn't `10e-10 kg/s`. If your model is small, it might be less than that. 

Do you have a link to the benchmark problem?

**Reply by batodon** [reply]
Yeah, sure. That's what I thought. I guess they were referring to mass flow rate instead of mass flux. Here is the link: https://doi.org/10.1016/j.ijrmms.2021.104879. Btw, the model is a 'unit' square. That makes the boundary 1m, i guess.

**Comment by cpgr** [comment, marked as answer]
I wonder if there is a typo in the paper. Assuming that I am reading it correctly, at zero strain it is basically a 1m x 1m block with `k=1e-19 m2`. If `mu = 1e-3 Pa.s`, `A = 1 m2`, `L = 1 m`, `Q = m / rho = 1e-10 / 1e3 = 1e-13`, then Darcy's law rearranged gives 

`Delta P = mu Q / k = 1e-3 x 1e-13 / 1e-19 = 1e3` which is basically what you get. 

For the published `Delta P = 1 MPa` at zero strain, `m = 1e-7 kg/s` (or `Q = 1e-10 m3/s`). So I wonder whether the authors just have the wrong units for their input flux?

**Reply by batodon** [reply]
Hi @cpgr, you’re absolutely right !. The authors got the units wrong; they should have reported `Q=1e-10 m3/s` instead of `1e-10 kg/s`. I just sent an email to their corresponding author. I’ll give you feedback on it soon. But even when I changed the flux to `1e-7 kg/s`, I had the same results as before, the gradient is extremely small, and it seems the (flux) change is not affecting the pore pressure change.  

Just to give you insight, I got the following results on my console with the same gradient (image/result) I posted previously:

 
![pp](https://user-images.githubusercontent.com/113298585/206600510-05548f9e-279b-4e51-a699-9290827574f2.PNG)

**Reply by batodon** [reply]
I think there is something fundamentally wrong with my input file. No matter the value I assign to the flux, I still get the same result...

**Reply by cpgr** [reply]
Can you tighten your tolerance? It hasn't actually done anything here, as the initial residual is smaller than your tolerance. Perhaps set `nl_abs_tol = 1e-12` for starters?

**Reply by batodon** [reply]
Great news! Now i'm getting there! I got results for 3D steady-state (i.e., cube) but yet to get a result for the 2D version (i.e., square). See below for the results i'm getting. 

Left=3D  Right= 2D

![3D(left) _2D(right)](https://user-images.githubusercontent.com/113298585/206611002-bf569842-242d-4efa-9a77-537bdf52d7a6.PNG)

**Reply by batodon** [reply]
Hi @cpgr, Just wanna let you know I got the results for the 2D case now. I attached below a comparison (both pressure and permeability) of the benchmark and my result. 

In all cases, Left = Benchmark Right = my results.

![pressure_compared](https://user-images.githubusercontent.com/113298585/206619208-f259acf0-c1b4-4703-996f-0f3c0da07b66.PNG)

![permeability_compared](https://user-images.githubusercontent.com/113298585/206619236-f344fa87-4ed8-4a02-b3bb-992ab634b9a2.PNG)

**Reply by batodon** [reply]
Btw, I got these results using `m=1e-7 kg/s`.

**Reply by cpgr** [reply]
OK, making progress! I guess the next step is to work out why your permeability is 3 times greater than the benchmark problem.

Also, it looks like the pressure is higher at the top corner than it is at the bottom corner in your results, when it should look more like the benchmark one I imagine.

**Reply by batodon** [reply]
Hello @cpgr ,
Here is a much better result. I’m finding it challenging to get the exact results from the benchmark. My pore-pressure seems to be half that of the benchmark. I’m guessing this might be due to the magnitude of my displacement. Not quiet sure because I assigned the proper displacement BCs. Also, my permeability looks a bit smaller (don't know if it is changing at all). Any suggestion? I’m still working on it. I attached the results here: 

my_results
![pp_disp_perm](https://user-images.githubusercontent.com/113298585/206769142-92b1d3ef-f28a-45ea-9710-f57b180197b3.PNG)

benchmark
![pp_disp_perm_benchmark](https://user-images.githubusercontent.com/113298585/206769172-be62af16-cd84-47ca-a0f9-6d1667d4add3.PNG)

**Reply by batodon** [reply]
Sorry, pls, ignore the previous comments about the displacements. I got the correct values now, but my pore pressure is still about half the correct value. My permeability has also not changed. It is still lesser than the benchmark value by a factor of 1.4.

**Reply by batodon** [reply]
Hello @cpgr,

So I received feedback from the corresponding author of the paper, and I just want to update you. In their literature, they used a density of `1kg/m3` instead of the usual density of water (i.e., `1e3 kg/m3`) that we’re familiar with and are using to compute the pressure gradient. That makes their mass flow rate equal to `1e10 kg/s`. After making the changes, I got an almost exact value (see attached), but the magnitude of my permeability is still a bit off, although it is the same in the X and Y direction. I’m still working on it. Thanks a lot!  

Benchmark= Left my_results =Right
![pp_compared](https://user-images.githubusercontent.com/113298585/207156557-82cbc6db-a950-4d25-9d79-1a60f96abd67.PNG)

![perm_compared](https://user-images.githubusercontent.com/113298585/207156592-0d443613-d010-4e42-a41f-4c997885593a.PNG)

**Reply by cpgr** [reply]
Well, I wasn't expecting that density! I'm glad you got to the bottom of that and that the pressure now looks pretty good. These benchmark problems are often tricky due to weird things like this.

I assume that your permeability is correct when there is no strain applied?

**Reply by batodon** [reply]
Thanks for the feedback, @cpgr . I believe (strongly) it is correct without the strain, because my input permeability is not changing at all. It makes me wonder (because the whole point of doing this simulation is to ensure that my strain affects the permeability). 
I’m trying to separate the permeability tensor into its components and see whether they will add up to give the exact magnitude in the benchmark, but I’m getting an error. I’ll open a ticket on that soon though.

**Reply by batodon** [reply]
@cpgr, if you have some time, could you please offer a second pair of eyes on the permeability material class that I’m using? Perhaps the material class is missing something very basic. Here is a link to it. https://github.com/iyevugah/ERGS/tree/main/src/materials. I’ll deeply be grateful. Thanks.

**Reply by batodon** [reply]
Hello @cpgr. If I rotate my strain through an angle of `0.04 radians`, the `I - M` tensor in my implementation/formulation changes, and I get a permeability of approximately `1.4e20` (similar to the benchmark study). My pore pressure, however, still remains unchanged. Without the rotation, `I-M` remains zero, and my permeability remains unaffected by the strain (i.e., the same as the intrinsic/input permeability). Any help will be appreciated. Thanks.  

![k_perm](https://user-images.githubusercontent.com/113298585/207467167-d61165b8-dd28-43ba-b856-651fbbd1284d.PNG)

```
Postprocessor Values:
+----------------+------------------------+----------------+----------------+
| time           | fracture_normal_strain | permeability   | pwater_pressure|
+----------------+------------------------+----------------+----------------+
|   0.000000e+00 |           0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|   1.000000e+00 |           4.262869e-05 |   1.416628e-20 |   1.079311e+07 |
+----------------+------------------------+----------------+----------------+

```

**Reply by cpgr** [reply]
I'll take a look when I get a chance

**Reply by batodon** [reply]
Hello @cpgr,

Some good news, finally. I think the problem is either with the precision of MOOSE compared to OGS or some trivial issue regarding my kernels. I will keep it brief!
First, it turns out that the square benchmark that I’m comparing my results against is a test to show that the permeability will not activate when the strain and the ‘fracture normal vector’ are orthogonal (i.e., there is no normal strain/`e_n`). For this to happen, the normal vector must not be rotated and could be in the x-direction (i.e., `1 0 0`) while the strain is perpendicular to it. Therefore, the input/intrinsic permeability will not change at all. Considering all these, it looks like my simulation is correct. Below, I compared both the ‘square’ test and my simulation. 
The only problem, however, is that assuming permeability does not change (i.e., at `1e-20`), the pore pressure should be `1e7 Pa` (i.e., `Delta P = mu Q / k = 1e-13/1e-20`), which isn’t the case here. This is why I think the error is either due to precision or some minor issue with my kernel. Please let me know.


square_test benchmark
![square_prj](https://user-images.githubusercontent.com/113298585/207774072-d36e64b7-e05c-49a8-9711-ea2a9cd7c5e6.PNG)

square_test myresults
![my_square_i](https://user-images.githubusercontent.com/113298585/207774098-43c32908-4203-4cfc-ad65-331fa0bda55f.PNG)

I now compared the actual benchmark 1 (i.e., a ‘strained’ cube instead of a square), where the strain is in the same direction as the normal vector (Z ), and the flow is orthogonal (X-direction), with my simulation in MOOSE. These results show that the simulation is on par with the benchmark. Like before, the pore pressure is a bit off!

strained_cube (Zstrain_Xflow) benchmark
![z_strain_x_flow](https://user-images.githubusercontent.com/113298585/207774213-788cc944-0912-4b62-99a3-a86e5ba1e2a0.PNG)

strained_cube (Zstrain_Xflow) myresult
![my_z_strain_x_flow](https://user-images.githubusercontent.com/113298585/207774283-15b45750-0a04-4383-849c-c80244a288ea.PNG)

**Reply by cpgr** [reply]
That's good investigating! One thing to check is that your density is exactly 1 kg/m3. Can you output that to check?

**Reply by cpgr** [reply]
By default, SimpleFluidProperties has a nonzero thermal expansion coefficient which can change the density even with a large bulk modulus.

**Reply by batodon** [reply]
This is what I have for the density:

```
Postprocessor Values:
+----------------+----------------+------------------------+----------------+----------------+
| time           | density        | fracture_normal_strain | permeability   | pwater_pressure|
+----------------+----------------+------------------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |           0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|   1.000000e+00 |   9.391932e-01 |          -4.285714e-05 |   1.000000e-20 |   1.074744e+07 |
+----------------+----------------+------------------------+----------------+----------------+
```

**Reply by batodon** [reply]
Thanks, @cpgr. It worked like a charm!!! :)

**Reply by cpgr** [reply]
I guess you already worked it out, but
```
[water]
  type = SimpleFluidProperties
  density0 = 1
  viscosity = 1e-3
  bulk_module = 1e9
  thermal_expansion = 0
[]
```
should work!

So does this mean you get the correct pressure now?

**Reply by batodon** [reply]
ok thanks a lot @cpgr. `1e9` looks way more better and exact! I got the exact pressures now!!! You're the best!

**Reply by wangshengwa** [reply]
Hi @batodon,

Great to know that you solved the problem.
Can you please comment what type of BCs did you use? Neumann or PorousFlowSink?
Cheers!

**Reply by batodon** [reply]
Hi @wangshengwa. PorousFlowSink acts as a source or sink (adding/removing fluid or heat through a boundary) depending on the ‘sign’. So, it can serve as a NeumannBC. In my case, since I’m injecting fluid from one side, I assigned PorousFlowSink to that side. I could easily assign NeumannBC, and it will work just fine. See: https://mooseframework.inl.gov/modules/porous_flow/boundaries.html. Thanks.
